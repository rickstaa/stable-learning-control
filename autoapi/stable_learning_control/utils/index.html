<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>stable_learning_control.utils &mdash; stable-learning-control 6.0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../../../_static/graphviz.css?v=eafc0fe6" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/modify.css?v=519ed47b" />

  
    <link rel="shortcut icon" href="../../../_static/favicon.ico"/>
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js?v=9497d378"></script>
        <script src="../../../_static/doctools.js?v=888ff710"></script>
        <script src="../../../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="stable_learning_control.utils.log_utils" href="log_utils/index.html" />
    <link rel="prev" title="stable_learning_control.disturbers.observation_random_noise_disturber" href="../disturbers/observation_random_noise_disturber/index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html">
            
              <img src="../../../_static/logo.svg" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                6.0.1
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../usage/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage/docker.html">Use with Docker</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage/algorithms.html">Available Agents</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage/running.html">Running Experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage/hyperparameter_tuning.html">Hyperparameter Tuning</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage/saving_and_loading.html">Experiment Outputs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage/plotting.html">Plotting Results</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage/eval_robustness.html">Evaluating Robustness</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage/benchmarks.html">Benchmarks</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Utilities</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../utils/loggers.html">Loggers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils/mpi.html">MPI Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils/run_utils.html">Run Utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils/plotter.html">Plotter</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils/testers.html">Policy testers</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Development</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../dev/contributing.html">Contribute to stable-learning-control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../dev/doc_dev.html">Build the documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../dev/license.html">License</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Documentation</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../../index.html">API Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../index.html">stable_learning_control</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="../index.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="../algos/index.html">stable_learning_control.algos</a></li>
<li class="toctree-l4"><a class="reference internal" href="../common/index.html">stable_learning_control.common</a></li>
<li class="toctree-l4"><a class="reference internal" href="../disturbers/index.html">stable_learning_control.disturbers</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">stable_learning_control.utils</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../index.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="../index.html#attributes">Attributes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../index.html#functions">Functions</a></li>
<li class="toctree-l3"><a class="reference internal" href="../index.html#package-contents">Package Contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Etc.</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../etc/acknowledgements.html">Acknowledgements</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">stable-learning-control</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="../index.html">stable_learning_control</a></li>
      <li class="breadcrumb-item active">stable_learning_control.utils</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/rickstaa/stable-learning-control/blob/main/docs/source/autoapi/stable_learning_control/utils/index.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-stable_learning_control.utils">
<span id="stable-learning-control-utils"></span><h1>stable_learning_control.utils<a class="headerlink" href="#module-stable_learning_control.utils" title="Permalink to this heading"></a></h1>
<p>Contains all the SLC utilities.</p>
<section id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Permalink to this heading"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="log_utils/index.html">stable_learning_control.utils.log_utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="mpi_utils/index.html">stable_learning_control.utils.mpi_utils</a></li>
</ul>
</div>
</section>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this heading"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="eval_robustness/index.html">stable_learning_control.utils.eval_robustness</a></li>
<li class="toctree-l1"><a class="reference internal" href="eval_utils/index.html">stable_learning_control.utils.eval_utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="gym_utils/index.html">stable_learning_control.utils.gym_utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="import_utils/index.html">stable_learning_control.utils.import_utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="plot/index.html">stable_learning_control.utils.plot</a></li>
<li class="toctree-l1"><a class="reference internal" href="run_entrypoint/index.html">stable_learning_control.utils.run_entrypoint</a></li>
<li class="toctree-l1"><a class="reference internal" href="run_utils/index.html">stable_learning_control.utils.run_utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="safer_eval_util/index.html">stable_learning_control.utils.safer_eval_util</a></li>
<li class="toctree-l1"><a class="reference internal" href="serialization_utils/index.html">stable_learning_control.utils.serialization_utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="test_policy/index.html">stable_learning_control.utils.test_policy</a></li>
</ul>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../disturbers/observation_random_noise_disturber/index.html" class="btn btn-neutral float-left" title="stable_learning_control.disturbers.observation_random_noise_disturber" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="log_utils/index.html" class="btn btn-neutral float-right" title="stable_learning_control.utils.log_utils" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Rick Staa.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>