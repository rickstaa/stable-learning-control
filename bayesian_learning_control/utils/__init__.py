"""Contains several utilities that can be used across all the BLC modules.
"""
