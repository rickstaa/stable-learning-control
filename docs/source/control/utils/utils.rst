=========
Utilities
=========

The :mlc:`Machine Learning Control <>` contains several Utilities for


.. toctree::
   :maxdepth: 2
   :caption: Control

   ../control
   ../algorithms/sac
   ../running
   ../saving_and_loading
   ../plotting
   ../bench/bench