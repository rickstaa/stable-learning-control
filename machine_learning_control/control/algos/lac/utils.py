from collections import OrderedDict

import numpy as np
import copy
import torch.nn as nn

from variant import (
    TRAIN_PARAMS,
    ENV_PARAMS,
)


def get_env_from_name(name, ENV_SEED):
    """Initializes the gym environment with the given name

    Args:
        name (str): The name of the gym environment you want to initialize.

    Returns:
        gym.Env: The gym environment.
    """
    if name == "oscillator":
        from envs.oscillator import oscillator as env

        env = env()
        env = env.unwrapped
    elif name == "Ex3_EKF":
        from envs.Ex3_EKF import Ex3_EKF as env

        env = env()
        env = env.unwrapped
    elif name == "Ex4_EKF":
        from envs.Ex4_EKF import Ex4_EKF as env

        env = env()
        env = env.unwrapped
    if ENV_SEED is not None:
        env.seed(ENV_SEED)
    return env


def evaluate_training_rollouts(paths):
    """Evaluate the performance of the policy in the training rollouts."""
    data = copy.deepcopy(paths)
    if len(data) < 1:
        return None
    try:
        diagnostics = OrderedDict(
            (
                ("return", np.mean([np.sum(path["rewards"]) for path in data])),
                ("length", np.mean([len(p["rewards"]) for p in data])),
            )
        )
    except KeyError:
        return
    [path.pop("rewards") for path in data]
    for key in data[0].keys():
        result = [np.mean(path[key]) for path in data]
        diagnostics.update({key: np.mean(result)})

    return diagnostics


def training_evaluation(env, policy):
    """Evaluates the performance of the current policy in
    several test rollouts.

    Args:
        env (gym.Env): The gym environment you want to use.
        policy (object): The current policy.

    Returns:
        [type]: [description]
    """
    # Retrieve action space bounds from env
    a_upperbound = env.action_space.high
    a_lowerbound = env.action_space.low

    # Training setting
    total_cost = []
    episode_length = []
    die_count = 0
    seed_average_cost = []

    # Perform roolouts to evaluate performance
    for i in range(TRAIN_PARAMS["num_of_evaluation_paths"]):
        cost = 0
        s = env.reset()
        for j in range(ENV_PARAMS["max_ep_steps"]):
            if ENV_PARAMS["eval_render"]:
                env.render()
            a = policy.choose_action(s, True)
            action = a_lowerbound + (a + 1.0) * (a_upperbound - a_lowerbound) / 2
            s_, r, done, _ = env.step(action)
            cost += r
            if j == ENV_PARAMS["max_ep_steps"] - 1:
                done = True
            s = s_
            if done:
                seed_average_cost.append(cost)
                episode_length.append(j)
                if j < ENV_PARAMS["max_ep_steps"] - 1:
                    die_count += 1
                break

    # Save evaluation results
    total_cost.append(np.mean(seed_average_cost))
    total_cost_mean = np.average(total_cost)
    average_length = np.average(episode_length)

    # Return evaluation results
    diagnostic = {
        "return": total_cost_mean,
        "average_length": average_length,
    }
    return diagnostic


def mlp(sizes, activation, output_activation=nn.Identity):
    """Create a multi-layered perceptron using pytorch.

    Args:
        sizes (list): The size of each of the layers.

        activation (torch.nn.modules.activation): The activation function used for the
            hidden layers.

        output_activation (torch.nn.modules.activation, optional): The activation
            function used for the output layers. Defaults to torch.nn.Identity.

    Returns:
        torch.nn.modules.container.Sequential: The multi-layered perceptron.
    """
    layers = []
    for j in range(len(sizes) - 1):
        act = activation if j < len(sizes) - 2 else output_activation
        layers += [nn.Linear(sizes[j], sizes[j + 1]), act()]
    return nn.Sequential(*layers)
