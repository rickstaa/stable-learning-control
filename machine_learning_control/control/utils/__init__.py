"""A set of common utilities used within the algorithm code.

This module is based on the utils module found in the the
`Spinning Up repository <https://github.com/openai/spinningup>`_.
"""
