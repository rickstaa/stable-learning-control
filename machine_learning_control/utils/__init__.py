"""Contains several utilities that can be used across all the MLC modules.
"""
