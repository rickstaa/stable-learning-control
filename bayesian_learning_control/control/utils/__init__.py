"""A set of common utilities used within the algorithm code.
"""

from bayesian_learning_control.control.utils.safer_eval import safer_eval
