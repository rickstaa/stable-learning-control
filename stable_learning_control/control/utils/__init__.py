"""A set of common utilities used within the algorithm code.
"""
from stable_learning_control.control.utils.safer_eval import safer_eval
