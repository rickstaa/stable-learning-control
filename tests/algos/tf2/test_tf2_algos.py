"""Checks if the Tensorflow algorithms in this package are still working as expected
when run on the CPU.

This is done by comparing the current output of the algorithm with a previously saved
snapshot. We do this so that we can be sure that the algorithm still gives a
deterministic result when seeded after a change in the code.
"""
import importlib

import gymnasium as gym  # noqa: F401
import pytest
from gymnasium.utils import seeding

ALGOS = [
    "stable_learning_control.control.algos.tf2.lac.lac",
    "stable_learning_control.control.algos.tf2.sac.sac",
]


@pytest.mark.parametrize("algo", ALGOS)
class TestTF2Algos:
    env = gym.make("Pendulum-v1")  # Used because it is a simple environment.

    # Seed the environment.
    env.np_random, seed = seeding.np_random(0)
    env.action_space.seed(0)
    env.observation_space.seed(0)

    def test_reproducibility(self, algo, snapshot):
        """Checks if the algorithm is still working as expected."""
        # Check if tensorflow is available.
        if not importlib.util.find_spec("tensorflow"):
            pytest.skip(
                "Tensorflow not available. Please install it using `pip install .[tf]`."
            )

        # Import the algorithm run function.
        run = getattr(importlib.import_module(algo), algo.split(".")[-1])

        # Run the algorithm.
        trained_policy, replay_buffer = run(
            lambda: self.env,
            seed=0,
            epochs=1,
            update_after=400,
            steps_per_epoch=800,
            logger_kwargs=dict(verbose=False),
            device="cpu",
        )

        # Test if the replay buffer is the same.
        assert replay_buffer == snapshot

        # Test if the actions returned by the policy are the same.
        for _ in range(5):
            action = trained_policy.get_action(self.env.observation_space.sample())
            assert action.numpy() == snapshot
