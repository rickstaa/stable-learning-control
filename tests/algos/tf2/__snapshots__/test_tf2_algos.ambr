# serializer version: 1
# name: TestTF2Algos.test_reproducibility[stable_learning_control.algos.tf2.lac.lac]
  ReplayBuffer(
    act_buf=array([[ 0.16709924],
         [ 0.8363838 ],
         [-1.243306  ],
         ...,
         [ 0.        ],
         [ 0.        ],
         [ 0.        ]], dtype=float32),
    done_buf=array([0., 0., 0., ..., 0., 0., 0.], dtype=float32),
    obs_buf=array([[ 0.6520163 ,  0.758205  , -0.46042657],
         [ 0.6469487 ,  0.76253355,  0.13329205],
         [ 0.61473   ,  0.78873765,  0.8306498 ],
         ...,
         [ 0.        ,  0.        ,  0.        ],
         [ 0.        ,  0.        ,  0.        ],
         [ 0.        ,  0.        ,  0.        ]], dtype=float32),
    obs_next_buf=array([[0.6469487 , 0.76253355, 0.13329205],
         [0.61473   , 0.78873765, 0.8306498 ],
         [0.5648556 , 0.82518977, 1.235707  ],
         ...,
         [0.        , 0.        , 0.        ],
         [0.        , 0.        , 0.        ],
         [0.        , 0.        , 0.        ]], dtype=float32),
    ptr=800,
    rew_buf=array([-0.76178324, -0.7545472 , -0.8963753 , ...,  0.        ,
          0.        ,  0.        ], dtype=float32),
    size=800,
  )
# ---
# name: TestTF2Algos.test_reproducibility[stable_learning_control.algos.tf2.lac.lac].1
  -1.6031608581542969
# ---
# name: TestTF2Algos.test_reproducibility[stable_learning_control.algos.tf2.lac.lac].2
  0.43915295600891113
# ---
# name: TestTF2Algos.test_reproducibility[stable_learning_control.algos.tf2.lac.lac].3
  0.33622145652770996
# ---
# name: TestTF2Algos.test_reproducibility[stable_learning_control.algos.tf2.lac.lac].4
  0.10656595230102539
# ---
# name: TestTF2Algos.test_reproducibility[stable_learning_control.algos.tf2.lac.lac].5
  -0.620768666267395
# ---
# name: TestTF2Algos.test_reproducibility[stable_learning_control.algos.tf2.latc.latc]
  ReplayBuffer(
    act_buf=array([[ 0.16709924],
         [ 0.8363838 ],
         [-1.243306  ],
         ...,
         [ 0.        ],
         [ 0.        ],
         [ 0.        ]], dtype=float32),
    done_buf=array([0., 0., 0., ..., 0., 0., 0.], dtype=float32),
    obs_buf=array([[ 0.6520163 ,  0.758205  , -0.46042657],
         [ 0.6469487 ,  0.76253355,  0.13329205],
         [ 0.61473   ,  0.78873765,  0.8306498 ],
         ...,
         [ 0.        ,  0.        ,  0.        ],
         [ 0.        ,  0.        ,  0.        ],
         [ 0.        ,  0.        ,  0.        ]], dtype=float32),
    obs_next_buf=array([[0.6469487 , 0.76253355, 0.13329205],
         [0.61473   , 0.78873765, 0.8306498 ],
         [0.5648556 , 0.82518977, 1.235707  ],
         ...,
         [0.        , 0.        , 0.        ],
         [0.        , 0.        , 0.        ],
         [0.        , 0.        , 0.        ]], dtype=float32),
    ptr=800,
    rew_buf=array([-0.76178324, -0.7545472 , -0.8963753 , ...,  0.        ,
          0.        ,  0.        ], dtype=float32),
    size=800,
  )
# ---
# name: TestTF2Algos.test_reproducibility[stable_learning_control.algos.tf2.latc.latc].1
  -1.5640698671340942
# ---
# name: TestTF2Algos.test_reproducibility[stable_learning_control.algos.tf2.latc.latc].2
  0.45661044120788574
# ---
# name: TestTF2Algos.test_reproducibility[stable_learning_control.algos.tf2.latc.latc].3
  0.3358030319213867
# ---
# name: TestTF2Algos.test_reproducibility[stable_learning_control.algos.tf2.latc.latc].4
  0.3512272834777832
# ---
# name: TestTF2Algos.test_reproducibility[stable_learning_control.algos.tf2.latc.latc].5
  -0.6032209396362305
# ---
# name: TestTF2Algos.test_reproducibility[stable_learning_control.algos.tf2.sac.sac]
  ReplayBuffer(
    act_buf=array([[ 0.28093982],
         [ 1.0383244 ],
         [-1.5125861 ],
         ...,
         [ 0.        ],
         [ 0.        ],
         [ 0.        ]], dtype=float32),
    done_buf=array([0., 0., 0., ..., 0., 0., 0.], dtype=float32),
    obs_buf=array([[ 0.6520163 ,  0.758205  , -0.46042657],
         [ 0.6462974 ,  0.7630856 ,  0.15036814],
         [ 0.61216897,  0.79072696,  0.878431  ],
         ...,
         [ 0.        ,  0.        ,  0.        ],
         [ 0.        ,  0.        ,  0.        ],
         [ 0.        ,  0.        ,  0.        ]], dtype=float32),
    obs_next_buf=array([[0.6462974 , 0.7630856 , 0.15036814],
         [0.61216897, 0.79072696, 0.878431  ],
         [0.5618093 , 0.82726675, 1.2445884 ],
         ...,
         [0.        , 0.        , 0.        ],
         [0.        , 0.        , 0.        ],
         [0.        , 0.        , 0.        ]], dtype=float32),
    ptr=800,
    rew_buf=array([-0.76183426, -0.7568918 , -0.911188  , ...,  0.        ,
          0.        ,  0.        ], dtype=float32),
    size=800,
  )
# ---
# name: TestTF2Algos.test_reproducibility[stable_learning_control.algos.tf2.sac.sac].1
  -0.8818840980529785
# ---
# name: TestTF2Algos.test_reproducibility[stable_learning_control.algos.tf2.sac.sac].2
  -0.5447466373443604
# ---
# name: TestTF2Algos.test_reproducibility[stable_learning_control.algos.tf2.sac.sac].3
  0.008646965026855469
# ---
# name: TestTF2Algos.test_reproducibility[stable_learning_control.algos.tf2.sac.sac].4
  1.228700876235962
# ---
# name: TestTF2Algos.test_reproducibility[stable_learning_control.algos.tf2.sac.sac].5
  -0.7967087030410767
# ---
